logging: uart_backend: add uncached async uart buffers support

JarmouniA opened the pull request:
When async UART is used & DCache is enabled, some SoCs, like STM32H7, require UART buffers to be placed in nocache memory to avoid cache coherence issues.

`__nocache` attribute will have no effect if `CONFIG_NOCACHE_MEMORY=n`

Fixes #87795

JarmouniA committed:
samples: subsys: logging: logger: add stm32h750b-dk conf

Enable the sample to run on stm32h750b-dk board with async
UART backend

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

pdgendt commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -237,8 +244,8 @@ const struct log_backend_api log_backend_uart_api = {
-				(.uart_dev = DEVICE_DT_GET(node_id),))                             \
-		.data = &lbu_data##__VA_ARGS__,                                                    \
+				(.uart_dev = DEVICE_DT_GET(node_id),)) .data =     \
+							   &lbu_data##__VA_ARGS__,                 \
```
No need to add this, clang-format isn't enforced.

pdgendt replied:
It's excluded in CI

https://github.com/zephyrproject-rtos/zephyr/blob/19c6240b6865bcb28e1d786d4dcadfb3a02067a0/.github/workflows/compliance.yml#L75

nordic-krch commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -225,8 +225,15 @@ const struct log_backend_api log_backend_uart_api = {
+#ifdef CONFIG_LOG_BACKEND_UART_ASYNC
+#define NOCACHE_ATTR __nocache
+#else
+#define ALIGN_BUF
```
is this used somewhere?

nordic-krch commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -225,8 +225,15 @@ const struct log_backend_api log_backend_uart_api = {
 	.format_set = format_set,
 };
 
+#ifdef CONFIG_LOG_BACKEND_UART_ASYNC
```
Putting buffer in __nocache will help but it will slow down the performance. Maybe it would be better to use STM guards for now. Other vendors may prefer to use cache line aligned memory and do the cache writeback which would have better performance. Nordic is using different appraoch. We have generic module used in drivers that detects if input buffer requires special action and uses memory pool to allocate buffer from a special location, do the copy and writeback if necessary.

JarmouniA committed:
logging: uart_backend: add uncached buffers support

When async UART is used & DCache is enabled, some SoCs, like STM32H7,
require UART buffers to be placed in nocache memory to avoid cache
coherence issues.

Fixes issue 87795

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

djiatsaf-st commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -225,8 +225,14 @@ const struct log_backend_api log_backend_uart_api = {
 	.format_set = format_set,
 };
 
+#if defined(CONFIG_LOG_BACKEND_UART_ASYNC) && defined(CONFIG_SOC_FAMILY_STM32) && defined(CONFIG_CPU_CORTEX_M7)
```
```suggestion
#if defined(CONFIG_LOG_BACKEND_UART_ASYNC) &&\ 
    defined(CONFIG_SOC_FAMILY_STM32) &&\
    defined(CONFIG_CPU_CORTEX_M7)
```

djiatsaf-st reviewed (commented):
LGTM

erwango commented on subsys/logging/backends/log_backend_uart.c:
```diff
@@ -225,8 +225,15 @@ const struct log_backend_api log_backend_uart_api = {
 	.format_set = format_set,
 };
 
+#if defined(CONFIG_LOG_BACKEND_UART_ASYNC) && defined(CONFIG_SOC_FAMILY_STM32) &&                  \
+	defined(CONFIG_CPU_CORTEX_M7)
```
> `select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU && CPU_HAS_ARM_MPU && CPU_HAS_DCACHE`

This actually encompasses STM32N6

erwango replied:
> We could make the check more flexible with CONFIG_SOC_FAMILY_STM32 && ARCH_HAS_NOCACHE_MEMORY_SUPPORT

It would probably be more future proof indeed.
Then what other families? Do you exclude them to remain on the safe side of what you've been able to test ?

erwango replied:
> , is it selected somewhere else?

https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/cortex_m/Kconfig#L71-L72
